### PR TITLE
Add splash attention integration from tokamax

### PR DESCRIPTION
**Pull request by suexu1025**
# Description
* Add splash attention tokamax integration. 
* Performance tested b/436334605
# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/layers/attention_op.py [hunk 1]
```diff
@@ -1049,11 +1052,13 @@ def tpu_flash_attention(
         segment_axis_names_kv = nn.logical_to_mesh_axes((BATCH, KV_LENGTH))
 
     if self.config.expert_shard_attention_option == EP_AS_CONTEXT:
-      axis_names_splash_kernel = nn.logical_to_mesh_axes(self.flash_axis_names_splash_kernel_ep)
+      # todo: needs support for tokamax splash attention kernel
+      # axis_names_splash_kernel = nn.logical_to_mesh_axes(self.flash_axis_names_splash_kernel_ep)
```
Comment (khatwanimohit) [inline src/MaxText/layers/attention_op.py]: nit: do we need these todos

File: src/MaxText/layers/attention_op.py [hunk 2]
```diff
@@ -36,6 +36,9 @@
 from flax import nnx
 from flax.linen import partitioning
 
+from tokamax._src.ops.experimental.tpu.splash_attention import splash_attention_kernel
```
Comment (khatwanimohit) [inline src/MaxText/layers/attention_op.py]: you might have to hide this import behing a jax version check. Similar to what we have done for megablox. 
This is needed because jax stable stack has not upgraded to 0.8.0 yet.

File: src/MaxText/layers/attention_op.py [hunk 3]
```diff
@@ -1094,6 +1099,23 @@ def tpu_flash_attention(
...
         v_layout=splash_attention_kernel.QKVLayout[global_v_layout],
+        attn_logits_soft_cap=attn_logits_soft_cap,
+        residual_checkpoint_name="context",
+        fwd_cost_estimate=pl.CostEstimate(
+            flops=self.config.cost_estimate_flops_fwd,
+            transcendentals=0,
+            bytes_accessed=0,
+        )
+        if self.config.cost_estimate_flops_fwd > 0
+        else None,
```
Comment (gobbleturk) [inline src/MaxText/layers/attention_op.py]: Curious does None by default use a nonzero cost estimate?

File: src/MaxText/layers/attention_op.py (hunk 1 above)
Comment (gobbleturk) [reply src/MaxText/layers/attention_op.py]: Rafi (and I by extension!) really don't like submitting commented out code. You should spell out what the TODO is and remove commented out code. Is axis_names_splash_kernel used to shard the mask? What is it used for?

File: src/MaxText/layers/attention_op.py (hunk 3 above)
Comment (suexu1025) [reply src/MaxText/layers/attention_op.py]: if pass None, it will use default tokamax cost estimation function.  llama 70b and 405b verifided this change doesn't affect the perf.

File: src/MaxText/layers/attention_op.py [hunk 4]
```diff
@@ -1122,35 +1144,42 @@ def tpu_flash_attention(
 
       mask &= ChunkedCausalMask(shape=(query.shape[2], key.shape[2]), chunk_size=self.chunk_attn_window_size)
 
-    # Create multi-head mask
-    multi_head_mask = splash_attention_mask.MultiHeadMask(masks=(mask,) * query.shape[1])
+    # Create mask - tokmax now just uses a single mask and assumes broadcast to all heads
+    single_head_mask = mask
+
+    max_logit_value, max_val = None, self.config.use_max_logit_estimate
```
Comment (gobbleturk) [inline src/MaxText/layers/attention_op.py]: Can we wrap this in a small function update_max_logit_value() or similar?

File: src/MaxText/layers/attention_op.py (hunk 3 above)
Comment (gobbleturk) [reply src/MaxText/layers/attention_op.py]: Can you include a comment in base.yml about this? Is the cost estimate seq^2 or seq^2/2 or similar?

File: src/MaxText/layers/attention_op.py [hunk 5]
```diff
@@ -22,20 +22,28 @@
...
+import jax
 from jax import lax
 from jax.ad_checkpoint import checkpoint_name
 from jax.experimental.pallas.ops.gpu import attention as gpu_pallas_attention
 from jax.experimental.pallas.ops.gpu import decode_attention as gpu_pallas_decode_attention
-from jax.experimental.pallas.ops.tpu.splash_attention import splash_attention_kernel
-from jax.experimental.pallas.ops.tpu.splash_attention import splash_attention_mask
+from jax.experimental import pallas as pl
+
+if jax.__version__ < "0.8.0":
```
Comment (khatwanimohit) [inline src/MaxText/layers/attention_op.py:34]: I don't think we need to protect this import under a condition

File: src/MaxText/layers/attention_op.py [hunk 6]
```diff
@@ -1081,21 +1091,53 @@ def tpu_flash_attention(
...
-        block_kv_dkv_compute=min(global_block_kv_dkv_compute, query.shape[2]),
-        block_q_dq=None if global_use_fused_bwd_kernel else min(global_block_q_dq, query.shape[2]),
-        block_kv_dq=None if global_use_fused_bwd_kernel else min(global_block_kv_dq, query.shape[2]),
-        use_fused_bwd_kernel=global_use_fused_bwd_kernel,
-        q_layout=splash_attention_kernel.QKVLayout[global_q_layout],
-        k_layout=splash_attention_kernel.QKVLayout[global_k_layout],
-        v_layout=splash_attention_kernel.QKVLayout[global_v_layout],
-    )
-
+    if jax.__version__ >= "0.8.0":
```
Comment (khatwanimohit) [inline src/MaxText/layers/attention_op.py]: for these calls and all the ones under this can we create a new flag in base.yml `use_tokamax_splash`

File: src/MaxText/layers/attention_op.py (hunk 5 above)
Comment (suexu1025) [reply src/MaxText/layers/attention_op.py:34]: otherwise, I got duplicated imports, as those share same module names

File: src/MaxText/layers/attention_op.py (hunk 6 above)
Comment (suexu1025) [reply src/MaxText/layers/attention_op.py]: each module imports on condition of jax version, so needs to be align with condition flag, jax._version_ here , otherwise will be compiler errors.

File: src/MaxText/layers/attention_op.py (hunk 5 above)
Comment (khatwanimohit) [reply src/MaxText/layers/attention_op.py:34]: ahh, maybe we can try to alias the tokamax splash imports

File: src/MaxText/layers/attention_op.py (hunk 3 above)
Comment (suexu1025) [reply src/MaxText/layers/attention_op.py]: chat offline, update the default to -1

File: src/MaxText/layers/attention_op.py (hunk 4 above)
Comment (suexu1025) [reply src/MaxText/layers/attention_op.py]: I cleaned up the max logit here.

File: src/MaxText/layers/attention_op.py (hunk 2 above)
Comment (suexu1025) [reply src/MaxText/layers/attention_op.py]: added jax version check

File: src/MaxText/layers/attention_op.py (hunk 1 above)
Comment (suexu1025) [reply src/MaxText/layers/attention_op.py]: removed comments

File: src/MaxText/layers/attention_op.py (hunk 5 above)
Comment (suexu1025) [reply src/MaxText/layers/attention_op.py:34]: for jax > 0.8.0, I don't think we need the flag to allow older splash. We will eventually migrate to tokamax after stable jax update to 0.8.0, wdyt?

File: generated_requirements/tpu-requirements.txt [hunk 7]
```diff
@@ -221,7 +221,7 @@ tensorflow>=2.19.1
 tensorstore>=0.1.78
 termcolor>=3.1.0
 tiktoken>=0.12.0
-tokamax>=0.0.3
+tokamax>=0.0.4
```
Comment (bvandermoon) [inline generated_requirements/tpu-requirements.txt:225]: Just to confirm - this file was built with `seed-env`, right?

File: src/MaxText/layers/attention_op.py (hunk 6 above)
Comment (bvandermoon) [reply src/MaxText/layers/attention_op.py]: @suexu1025 you are saying there is no need for a config flag here, right? If flash attention is used, then this should be used in all cases?

Comment (suexu1025) [reply src/MaxText/layers/attention_op.py]: yes, this will be for jax >= 0.8.0,  don't want to maintain and keep 2 version of pallas splash kernel.

File: src/MaxText/layers/attention_op.py [hunk 8]
```diff
@@ -1081,21 +1089,53 @@ def tpu_flash_attention(
...
+      sa_config = splash_attention_kernel.SplashConfig(
+          block_q=min(global_block_q, query.shape[2]),
```
Comment (gobbleturk) [inline src/MaxText/layers/attention_op.py]: Can you create a subfunction to create sa_config, e.g. get_sa_config(...):

File: generated_requirements/tpu-requirements.txt (hunk 7 above)
Comment (suexu1025) [reply generated_requirements/tpu-requirements.txt:225]: updated file was generated by seed-env

File: src/MaxText/layers/attention_op.py (hunk 8 above)
Comment (suexu1025) [reply src/MaxText/layers/attention_op.py]: updated, and can you check?

**Comment by bvandermoon** [comment]
> LGTM, tokamax 0.4.0 can't be added to the JAX AI Image yet, right? Since they aren't on the latest JAX yet

Or is it also needed in [requirements.txt](https://github.com/AI-Hypercomputer/maxtext/blob/58dba0a78f8ca34e20a778892b2ce4409f2067d8/requirements.txt#L40)?

File: src/MaxText/layers/attention_op.py (hunk 6 above)
Comment (suexu1025) [reply src/MaxText/layers/attention_op.py]: I can follow a PR to add a flag to use old attention once the cluster issue is resolved.

File: src/MaxText/layers/attention_op.py [hunk 9]
```diff
@@ -22,20 +22,28 @@
...
-from jax.experimental.pallas.ops.tpu.splash_attention import splash_attention_kernel
-from jax.experimental.pallas.ops.tpu.splash_attention import splash_attention_mask
+from jax.experimental import pallas as pl
 from jax.sharding import Mesh
-import jax
 import jax.numpy as jnp
 
+if jax.__version__ < "0.8.0":
+  from jax.experimental.pallas.ops.tpu.splash_attention import splash_attention_kernel
+  from jax.experimental.pallas.ops.tpu.splash_attention import splash_attention_mask
```
Comment (parambole) [inline src/MaxText/layers/attention_op.py:36]: Do we still need this condition as tokamax==0.4.0 requires JAX to be at 0.8.0
